File: app/src/main/java/com/tcssol/expensetracker/Model/Expenses.java
package com.tcssol.expensetracker.Model;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;


import com.tcssol.expensetracker.Utils.Converters;

import java.time.LocalDate;
@Entity(tableName="expenses_table")
public class Expenses {
    @PrimaryKey(autoGenerate=true)
    private long id;
    @ColumnInfo(name="date_created")
    private LocalDate dateCreated;
    private String category;
    @ColumnInfo(name="sub_category")
    private String subCategory;
    private String mode;
    private double amount;
    private boolean type;

    @Ignore
    public Expenses(String category,String subCategory){
        this.category=category;
        this.subCategory=subCategory;
    }

    public Expenses(LocalDate dateCreated, String category, String subCategory,String mode, double amount, boolean type) {
        this.dateCreated = dateCreated;
        this.category = category;
        this.subCategory = subCategory;
        this.mode=mode;
        this.amount = amount;
        this.type = type;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getId() {
        return id;
    }

    public LocalDate getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(LocalDate dateCreated) {
        this.dateCreated = dateCreated;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getSubCategory() {
        return subCategory;
    }

    public void setSubCategory(String subCategory) {
        this.subCategory = subCategory;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public boolean isType() {
        return type;
    }

    public void setType(boolean type) {
        this.type = type;
    }

    public String getMode() {
        return mode;
    }

    public void setMode(String mode) {
        this.mode = mode;
    }

    public static String toCsvFormat(Expenses expenses){
        String typeString=String.valueOf(expenses.isType());
        String temp= String.format("%d,%s,%s,%s,%s,%s,%s\n", expenses.getId(), expenses.getCategory(), expenses.getSubCategory(),expenses.getMode(), Converters.toString(expenses.getDateCreated()), typeString, expenses.getAmount());
        return temp;
    }
    public static String toTxtFormat(Expenses expenses){
        String typeString=String.valueOf(expenses.isType());
        String temp= String.format("%d\t%s\t%s\t%s\t%s\t%s\t%s\n", expenses.getId(), expenses.getCategory(), expenses.getSubCategory(),expenses.getMode(), Converters.toString(expenses.getDateCreated()), typeString, expenses.getAmount());
        return temp;
    }
}

File: app/src/main/java/com/tcssol/expensetracker/Model/Expenses.kt
package com.tcssol.expensetracker.Model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.tcssol.expensetracker.Utils.Converters
import java.time.LocalDate

@Entity(tableName = "expenses_table")
data class Expenses(
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0,
    
    @ColumnInfo(name = "date_created")
    var dateCreated: LocalDate,
    
    var category: String,
    
    @ColumnInfo(name = "sub_category")
    var subCategory: String,
    
    var mode: String,
    
    var amount: Double,
    
    var type: Boolean,
    
    @ColumnInfo(name = "note")
    var note: String? = null
) {
    // Secondary constructors for backward compatibility & easy creation
    @Ignore
    constructor(category: String, subCategory: String) : this(
        dateCreated = LocalDate.now(),
        category = category,
        subCategory = subCategory,
        mode = "Cash",
        amount = 0.0,
        type = false
    )

    @Ignore
    constructor(
        dateCreated: LocalDate,
        category: String,
        subCategory: String,
        mode: String,
        amount: Double,
        type: Boolean
    ) : this(
        id = 0,
        dateCreated = dateCreated,
        category = category,
        subCategory = subCategory,
        mode = mode,
        amount = amount,
        type = type,
        note = null
    )

    companion object {
        @JvmStatic
        fun toCsvFormat(expenses: Expenses): String {
            val typeString = expenses.type.toString()
            val noteStr = expenses.note?.replace(",", ";") ?: ""
            return String.format(
                "%d,%s,%s,%s,%s,%s,%s,%s\n",
                expenses.id,
                expenses.category,
                expenses.subCategory,
                expenses.mode,
                Converters.toString(expenses.dateCreated),
                typeString,
                expenses.amount,
                noteStr
            )
        }

        @JvmStatic
        fun toTxtFormat(expenses: Expenses): String {
            val typeString = expenses.type.toString()
            val noteStr = expenses.note ?: ""
            return String.format(
                "%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n",
                expenses.id,
                expenses.category,
                expenses.subCategory,
                expenses.mode,
                Converters.toString(expenses.dateCreated),
                typeString,
                expenses.amount,
                noteStr
            )
        }
    }
}
